### PR TITLE
Sinatra webhook

### PR DESCRIPTION
**Pull request by josephwegner**
## What is this
This _should_ add support for webhook-initiated jobs. Webhooks would allow us to start jobs immediately when an event happens, rather than polling for changes every N minutes.

Internally the web server is a very basic Sinatra app. Every time the user adds a webhook job, we register a new route with sinatra - we listen to GET, PUT, and POST requests (all have an identical handler).

The user will be able to define the route for the webhook to whatever they want. This should also probably have some concept of authentication (HTTP basic, or maybe a secret GET param key).

## How the user initiates it

Webhook jobs are created by using `webhook` instead of `every`. `webhook` should be called with a path and a block - the block will get called whenever a request is made to that path, and it will execute in the sinatra request's scope. That's helpful, because things like `params` can be accessed right there.

```ruby
job do
  
  webhook '/path' do
    if params['secret_key'] == 'abcdef'
      params['data']
    else
      false
    end
  end

  step 'test' do
    puts 'test'
  end
end
```

### Starting with rackup, thin, etc.
config.ru
```ruby
require 'pushpop'

Pushpop.load_jobs # This loads everything in /jobs... The user can pass a specific filename here too
Pushpop.schedule # This schedules clock jobs... this can probably be moved into start_clock, actually
Pushpop.start_clock # Starts the actual clockwork manager in a new thread

run Pushpop.web.app # Starts the sinatra app
```

Then just run:
`rackup`
or
`thin start`

### Starting with unicorn

`config.ru`
```ruby
run Pushpop.web.app
```

`unicorn.rb`
``` ruby
require 'pushpop'

worker_processes 2 # set to whatever you want

Pushpop.load_jobs # This loads everything in /jobs... The user can pass a specific filename here too
Pushpop.schedule # This schedules clock jobs... this can probably be moved into start_clock, actually
Pushpop.start_clock # Starts the actual clockwork manager in a new thread
```

Then run:
`unicorn -c unicorn.rb`

### Triggering
Webhooks can be hit very simply

```bash
cURL localhost:4567/path
```
*PUT, POST, GET are all supported*

## Scary Threads
I don't know a whole ton about threading in ruby, so this makes me nervous... I had to move Clockwork into its own separate thread. If you know anything about threading in ruby (also, thread limits on Heroku), please let me know if I'm being dumb.

## Dreaming
- We could probably expose the sinatra app in a fairly simple way, so that users could build more things on the web server. This would make building pushpop-powered dashboards (similar to PingPong) fairly easy

### CODE REVIEW

**Comment by hex337** [comment]
So, for dealing with the starting of this, we should probably use [profiles](https://devcenter.heroku.com/articles/procfile).

From the api standpoint, I think that's fine for now. Accepting GET params on them will be good, and for authentication, basic http should be fine.

**Comment by hex337** [comment]
On the Threading side... we'll probably just have to experiment and see what combo works.

**Comment by josephwegner** [comment]
`params` support added. I updated the OP

tl;dr: `webhook` gets a block now, which is eval'd in the sinatra request's scope.

**Comment by hex337** [comment]
So, one thing to consider, maybe we should make this web part a plugin, since there are instances where you would want to run pushpop as a headless process (pingpong for example).

**Comment by josephwegner** [comment]
@hex337 Yeah - by default, pushpop is still headless. In fact, this whole thing (so far) is still backwards compatible.

Starting pushpop the normal way (`pushpop jobs:run`) will only run Clockwork jobs (it's currently broken, because of Clockwork being in a thread - I need to fix that). Similarly, I believe PingPong manually starts Pushpop via an `app.rb` calling `Pushpop.schedule` and `Clockwork.manager.run` - that will continue to work, although there is now a utility function inside of Pushpop for starting clock.

@hex337 I didn't realize, though, that Pushpop had plugin support. If headless support is already built in, would you still prefer to see this as a plugin? It feels like webhooks could be a pretty core feature to PushPop, so I'd like to see it in the main codebase, but I definitely see both sides.

**Comment by josephwegner** [comment]
Changed a bunch of stuff about how threads are managed and how the app starts internally... This makes stuff like unicorn work much better!

I also updated the OP with all of the new info.

**Comment by hex337** [comment]
Nice, those changes look good to me.

**Comment by josephwegner** [comment]
Added specs.

**Comment by josephwegner** [comment]
Updated README.

@hex337 from my perspective, this is good to release. If you have more thoughts/comments please let me know - I'd like to get this rock solid before releasing.

Also, I wouldn't mind you trying to get a webhook running locally just to prove that this is usable :smile:

**Comment by hex337** [comment]
I'll try to test this out today.

**Comment by hex337** [comment]
This looks good to me. I didn't do any extensive testing, but the basics work!
